src/main/java/com/mongodb/kafka/connect/sink/MongoSinkTask.java

```diff
@@ -186,6 +185,11 @@ private void bulkWriteBatch(final List<MongoProcessedSinkRecordData> batch) {
           writeModels.size(),
           namespace.getFullName());
       handleMongoException(config, e);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      if (!config.tolerateErrors()) {
+        throw new DataException("Rate limiting was interrupted", e);
```

Comment:
With the current approach when `tolerateErrors` is true, we finish `MongoSinkTask.put` by storing all documents but effectively ignoring the rate limit. This is probably not what we want.

I believe our reaction to interrupt signals should not depend on whether we are configured to tolerate errors or not, and this reaction could be either of the following
1. re-set interrupt flag, stop storing the docs and throw an non-retriable exception
2. remember to re-set interrupt flag once we store all the docs (or fail to store them for some reason), continue storing the docs without re-setting the flag so that rate limiting is respected, re-set the flag at the end.

The approach 2 is much harder to implement, but what is much more important, we probably don't actually know how `MongoClient.bulkWrite` behaves. And we can only reliably implement 2 if we are guaranteed that `MongoClient.bulkWrite` finishes its work completely.

Reply:
Think option 1 seems best. Will update.

Reply:
Done

Reply:
I also think that the option 1 is better.